### PR TITLE
Select GPU variant of eks optimized ami for nvidia & neuron

### PR DESCRIPTION
**Pull request by JacobGabrielson**
**1. Issue, if available:**

https://github.com/awslabs/karpenter/issues/683

**2. Description of changes:**

Uses the "-gpu" variant of eks optimized AMI if there are any nvidia instances in the list of instance types

Prelim test results:

Applied to cluster w/ no nvidia GPU instances:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nvidia-smi
spec:
  restartPolicy: OnFailure
  containers:
  - name: nvidia-smi
    image: nvidia/cuda:latest
    args:
    - "nvidia-smi"
    resources:
      limits:
        nvidia.com/gpu: 4
```

karpenter-controller logs:

```
2021-09-16T21:57:54.552884348Z 2021-09-16T21:57:54.552Z	INFO	controller.allocation.provisioner/default	Starting provisioning loop	{"commit": "569c3bc"}
2021-09-16T21:57:54.552932259Z 2021-09-16T21:57:54.552Z	INFO	controller.allocation.provisioner/default	Waiting to batch additional pods	{"commit": "569c3bc"}
2021-09-16T21:57:55.689190601Z 2021-09-16T21:57:55.689Z	INFO	controller.allocation.provisioner/default	Found 1 provisionable pods	{"commit": "569c3bc"}
2021-09-16T21:57:56.602742898Z 2021-09-16T21:57:56.602Z	DEBUG	controller.allocation.provisioner/default	Discovered 309 EC2 instance types	{"commit": "569c3bc"}
2021-09-16T21:57:56.604835556Z 2021-09-16T21:57:56.604Z	DEBUG	controller.allocation.provisioner/default	Excluding instance type t3.nano because there are not enough resources for kubelet and system overhead	{"commit": "569c3bc"}
2021-09-16T21:57:56.605042600Z 2021-09-16T21:57:56.604Z	DEBUG	controller.allocation.provisioner/default	Excluding instance type t3a.nano because there are not enough resources for kubelet and system overhead	{"commit": "569c3bc"}
2021-09-16T21:57:56.606514867Z 2021-09-16T21:57:56.606Z	INFO	controller.allocation.provisioner/default	Computed packing for 1 pod(s) with instance type option(s) [g2.8xlarge g4dn.12xlarge p3.8xlarge g3.16xlarge p2.8xlarge p3.16xlarge p3dn.24xlarge p4d.24xlarge p2.16xlarge]	{"commit": "569c3bc"}
2021-09-16T21:57:56.850586631Z 2021-09-16T21:57:56.850Z	DEBUG	controller.allocation.provisioner/default	Discovered 3 subnets for cluster jacob-karpenter-demo	{"commit": "569c3bc"}
2021-09-16T21:57:56.857076091Z 2021-09-16T21:57:56.856Z	DEBUG	controller.allocation.provisioner/default	Discovered kubernetes version 1.20	{"commit": "569c3bc"}
2021-09-16T21:57:56.924673505Z 2021-09-16T21:57:56.924Z	DEBUG	controller.allocation.provisioner/default	Discovered ami ami-0b1f0fc1fb3651d9f for query /aws/service/eks/optimized-ami/1.20/amazon-linux-2-gpu/recommended/image_id	{"commit": "569c3bc"}
2021-09-16T21:57:57.131200476Z 2021-09-16T21:57:57.130Z	DEBUG	controller.allocation.provisioner/default	Discovered 1 security groups for cluster jacob-karpenter-demo	{"commit": "569c3bc"}
2021-09-16T21:57:57.131240937Z 2021-09-16T21:57:57.131Z	DEBUG	controller.allocation.provisioner/default	Discovered caBundle, length 1066	{"commit": "569c3bc"}
2021-09-16T21:57:57.251842641Z 2021-09-16T21:57:57.251Z	DEBUG	controller.allocation.provisioner/default	Created launch template, Karpenter-jacob-karpenter-demo-15956394333737086452	{"commit": "569c3bc"}
2021-09-16T21:57:58.939682393Z 2021-09-16T21:57:58.939Z	INFO	controller.allocation.provisioner/default	Launched instance: i-0f9d2818ce653bdb0, type: g2.8xlarge, zone: us-west-2b, hostname: ip-192-168-132-145.us-west-2.compute.internal	{"commit": "569c3bc"}
2021-09-16T21:57:58.984805548Z 2021-09-16T21:57:58.984Z	INFO	controller.allocation.provisioner/default	Bound 1 pod(s) to node ip-192-168-132-145.us-west-2.compute.internal	{"commit": "569c3bc"}
```

`nvidia-smi` logs:

```
kubectl logs nvidia-smi
Thu Sep 16 22:43:05 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.73.01    Driver Version: 460.73.01    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GRID K520           On   | 00000000:00:03.0 N/A |                  N/A |
| N/A   32C    P8    N/A /  N/A |      0MiB /  4037MiB |     N/A      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  GRID K520           On   | 00000000:00:04.0 N/A |                  N/A |
| N/A   34C    P8    N/A /  N/A |      0MiB /  4037MiB |     N/A      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  GRID K520           On   | 00000000:00:05.0 N/A |                  N/A |
| N/A   34C    P8    N/A /  N/A |      0MiB /  4037MiB |     N/A      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  GRID K520           On   | 00000000:00:06.0 N/A |                  N/A |
| N/A   37C    P8    N/A /  N/A |      0MiB /  4037MiB |     N/A      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

**3. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/cloudprovider/aws/ami.go [hunk 1]
```diff
@@ -44,17 +45,22 @@ func NewAMIProvider(ssm ssmiface.SSMAPI, clientSet *kubernetes.Clientset) *AMIPr
...
 	}
-	var architectureSuffix string
+	var amiNameSuffix string
 	if *constraints.Architecture == v1alpha3.ArchitectureArm64 {
-		architectureSuffix = "-arm64"
+		amiNameSuffix = "-arm64"
 	}
-	// TODO: support for the "amazon-linux-2-gpu" AMI for nvidia use cases
-	name := fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2%s/recommended/image_id", version, architectureSuffix)
+	if AnyNvidiaGPUs(instanceTypes) {
```
Comment (bwagner5) [inline pkg/cloudprovider/aws/ami.go:57]: should it also check for AWS Neurons? 

It looks like the gpu ami ships with [AWS Neuron Container Runtime](https://docs.aws.amazon.com/eks/latest/userguide/eks-optimized-ami.html#gpu-ami)

Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/ami.go:57]: @bwagner5 I don't *think* so - the docs make it sound like this is (despite the name) only needed for nvidia gpus. But I would love a second opinion :-)

Comment (mikestef9) [reply pkg/cloudprovider/aws/ami.go:57]: It should also check for Inferentia instances. The EKS accelerated AMI works for Inf1 and Nvidia GPU instance types. It selects the correct container runtime by looking at the instance type through IMDS on startup

Comment (ellistarn) [reply pkg/cloudprovider/aws/ami.go:57]: Does it hard code instance types? Will this break for new instances or is there something smarter?

File: pkg/cloudprovider/aws/ami.go [hunk 2]
```diff
@@ -44,17 +45,22 @@ func NewAMIProvider(ssm ssmiface.SSMAPI, clientSet *kubernetes.Clientset) *AMIPr
...
 	}
-	// TODO: support for the "amazon-linux-2-gpu" AMI for nvidia use cases
-	name := fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2%s/recommended/image_id", version, architectureSuffix)
+	if NeedsGPUAmi(instanceTypes) {
+		if amiNameSuffix != "" {
+			return "", fmt.Errorf("no amazon-linux-2 ami available for both nvidia gpus and arm64 cpus")
+		}
+		amiNameSuffix = "-gpu"
+	}
+	name := fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2%s/recommended/image_id", version, amiNameSuffix)
```
Comment (ellistarn) [inline pkg/cloudprovider/aws/ami.go:63]: Can you move all of this ami name logic into a separate function?

File: pkg/cloudprovider/aws/instancetype.go [hunk 3]
```diff
@@ -83,6 +84,29 @@ func (i *InstanceType) NvidiaGPUs() *resource.Quantity {
...
+		if !i.NvidiaGPUs().IsZero() || !i.AWSNeurons().IsZero() {
+			return true
+		}
+	}
+	return false
+}
+
+// NeedsDocker returns true if the instance type is unable to use
+// conatinerd directly
+func NeedsDocker(is []cloudprovider.InstanceType) bool {
```
Comment (ellistarn) [inline pkg/cloudprovider/aws/instancetype.go:98]: IMO these functions are better encapsulated by launchtemplate.go. Until now, this file was just a pure implementation of the cloudprovider.InstanceType interface. If you move them, I'd make the methods private, too.

File: pkg/cloudprovider/aws/launchtemplate.go [hunk 4]
```diff
@@ -191,13 +192,18 @@ func (p *LaunchTemplateProvider) getSecurityGroupIds(ctx context.Context, provis
 	return securityGroupIds, nil
 }
 
-func (p *LaunchTemplateProvider) getUserData(ctx context.Context, provisioner *v1alpha3.Provisioner, constraints *Constraints) (string, error) {
+func (p *LaunchTemplateProvider) getUserData(ctx context.Context, provisioner *v1alpha3.Provisioner, constraints *Constraints, instanceTypes []cloudprovider.InstanceType) (string, error) {
+	var containerRuntimeArg string
+	if !NeedsDocker(instanceTypes) {
```
Comment (ellistarn) [inline pkg/cloudprovider/aws/launchtemplate.go:197]: How hard is it to remove the docker dependency?

**Comment by ellistarn** [comment]
Does Karpenter play well w/ the nvidia driver installer? How long does it take to bring up a new gpu node?

File: pkg/cloudprovider/aws/launchtemplate.go (hunk 4 above)
Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/launchtemplate.go:197]: Sorry, I do not understand the question, can you clarify?

**Comment by JacobGabrielson** [comment]
> Does Karpenter play well w/ the nvidia driver installer? How long does it take to bring up a new gpu node?

It's a good question - I'm not sure yet.

File: pkg/cloudprovider/aws/instancetype.go (hunk 3 above)
Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/instancetype.go:98]: Sounds good.

File: pkg/cloudprovider/aws/ami.go (hunk 2 above)
Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/ami.go:63]: Aye, aye, cap'n

File: pkg/cloudprovider/aws/ami.go [hunk 5]
```diff
@@ -45,7 +45,7 @@ func NewAMIProvider(ssm ssmiface.SSMAPI, clientSet *kubernetes.Clientset) *AMIPr
 	}
 }
 
-func (p *AMIProvider) Get(ctx context.Context, constraints *Constraints, instanceTypes []cloudprovider.InstanceType) (string, error) {
+func (p *AMIProvider) getSSMParameter(ctx context.Context, constraints *Constraints, instanceTypes []cloudprovider.InstanceType) (string, error) {
```
Comment (ellistarn) [inline pkg/cloudprovider/aws/ami.go:48]: completely optional: I might move this function below `Get()`. I tend to follow a mixed pattern of DFS and "bigger concepts to the top"

File: pkg/cloudprovider/aws/launchtemplate.go (hunk 4 above)
Comment (ellistarn) [reply pkg/cloudprovider/aws/launchtemplate.go]: How hard is it to use containerd for GPUs? Happy if the answer is "out of scope for now"

**Review by ellistarn — COMMENTED** [review]
LGTM. Is this working or still WIP?

File: pkg/cloudprovider/aws/launchtemplate.go (hunk 4 above)
Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/launchtemplate.go]: My understanding is, it's not hard. The EKS optimized AMI does already support it for Nvidia GPUs, just not for Neurons.

File: pkg/cloudprovider/aws/ami.go (hunk 1 above)
Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/ami.go]: @ellistarn this code doesn't hard-code instance types

Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/ami.go]: > should it also check for AWS Neurons?
> 
> It looks like the gpu ami ships with [AWS Neuron Container Runtime](https://docs.aws.amazon.com/eks/latest/userguide/eks-optimized-ami.html#gpu-ami)

done

File: pkg/cloudprovider/aws/ami.go (hunk 2 above)
Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/ami.go]: Done

File: pkg/cloudprovider/aws/instancetype.go (hunk 3 above)
Comment (JacobGabrielson) [reply pkg/cloudprovider/aws/instancetype.go]: Done

**Comment by JacobGabrielson** [comment]
> LGTM. Is this working or still WIP?

I'd like to check it in. I'm still testing, and it seems likely there is some kind of race going on with the nvidia daemonset, but I think this code is arguably a step in the right direction (gives things a chance of working), and I'm getting tired of rebasing all the time :-)

**Review by bwagner5 — APPROVED** [review]
lgtm!

File: pkg/cloudprovider/aws/launchtemplate.go [hunk 6]
```diff
@@ -149,6 +145,28 @@ func (p *LaunchTemplateProvider) ensureLaunchTemplate(ctx context.Context, optio
 	return launchTemplate, nil
 }
 
+func needsGPUAmi(is []cloudprovider.InstanceType) bool {
```
Comment (ellistarn) [inline pkg/cloudprovider/aws/launchtemplate.go:148]: i might've misled you earlier. Should these live in ami.go (where they're used)?

**Review by ellistarn — APPROVED** [review]
minor, optional

**Pull request merged**